### PR TITLE
Add scheduler for cosine in linear envelope.

### PR DESCRIPTION
**Pull request by dwadden**
This didn't prove particularly effective for annealing, but might as well have it as an option.

### CODE REVIEW

File: olmo/optim.py [hunk 1]
```diff
@@ -788,6 +789,22 @@ def get_lr(self, initial_lr: float, step: int, max_steps: int) -> float:
         return initial_lr
 
 
+@dataclass
+class CosLinearEnvelope(Scheduler):
+    "Pointwise product of cosine schedule and linear decay; useful during annealing."
+    t_max: Optional[int] = None
+
+    def get_lr(self, initial_lr: float, step: int, max_steps: int) -> float:
+        eta_min = 0.0
```
Comment (2015aroras) [inline olmo/optim.py:798]: Consider setting this to `initial_lr * alpha_f`, like with the cosine and linear schedulers. Warm up would probably be nice too.

Comment (2015aroras) [reply olmo/optim.py:798]: If you use `alpha_f`, you may need to update the logic for computing the envelope.

Comment (dwadden) [reply olmo/optim.py]: Changed, lmk if this looks good.

**Review by 2015aroras — APPROVED** [review]
LGTM!

**Pull request merged**